x-pack/filebeat/input/awscloudwatch/input.go

```diff
@@ -0,0 +1,360 @@
+	}
+
+	logger.Debug("awscloudwatch input config = ", config)
+	logGroupName, regionName, err := parseARN(config.LogGroupARN)
+	if err != nil {
+		return nil, errors.Wrap(err, "parse log group ARN failed")
+	}
+
+	config.LogGroup = logGroupName
+	config.RegionName = regionName
+
+	awsConfig, err := awscommon.GetAWSCredentials(config.AwsConfig)
+	if err != nil {
+		return nil, errors.Wrap(err, "getAWSCredentials failed")
+	}
+	awsConfig.Region = regionName
+
+	closeChannel := make(chan struct{})
+	in := &awsCloudWatchInput{
+		config:      config,
+		awsConfig:   awsConfig,
+		logger:      logger,
+		close:       closeChannel,
+		inputCtx:    &channelContext{closeChannel},
+		prevEndTime: int64(0),
+	}
+
+	// Build outlet for events.
+	in.outlet, err = connector.ConnectWith(cfg, beat.ClientConfig{
+		ACKEvents: func(privates []interface{}) {
+			for _, private := range privates {
+				if cwContext, ok := private.(*cwContext); ok {
+					cwContext.done()
+				}
+			}
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	in.logger.Info("Initialized AWS CloudWatch input.")
+	return in, nil
+}
+
+// Run runs the input
+func (in *awsCloudWatchInput) Run() {
+	in.workerOnce.Do(func() {
+		in.workerWg.Add(1)
+		go func() {
+			in.logger.Infof("awscloudwatch input worker for log group: '%v' has started", in.config.LogGroup)
+			defer in.logger.Infof("awscloudwatch input worker for log group '%v' has stopped.", in.config.LogGroup)
+			defer in.workerWg.Done()
+			in.run()
+		}()
+	})
+}
+
+func (in *awsCloudWatchInput) run() {
+	cwConfig := awscommon.EnrichAWSConfigWithEndpoint(in.config.AwsConfig.Endpoint, "cloudwatchlogs", in.config.RegionName, in.awsConfig)
+	svc := cloudwatchlogs.New(cwConfig)
+	for in.inputCtx.Err() == nil {
+		err := in.getLogEventsFromCloudWatch(svc)
+		if err != nil {
+			if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == awssdk.ErrCodeRequestCanceled {
+				continue
+			}
+			in.logger.Error("getLogEventsFromCloudWatch failed: ", err)
+			continue
+		}
+
+		in.logger.Debugf("sleeping for %v before checking new logs", in.config.ScanFrequency)
+		time.Sleep(in.config.ScanFrequency)
+		in.logger.Debug("done sleeping")
+	}
+}
+
+func parseARN(logGroupARN string) (string, string, error) {
+	arnParsed, err := arn.Parse(logGroupARN)
+	if err != nil {
+		return "", "", errors.Errorf("error Parse arn %s: %v", logGroupARN, err)
+	}
+
+	if strings.Contains(arnParsed.Resource, ":") {
+		resourceARNSplit := strings.Split(arnParsed.Resource, ":")
+		if len(resourceARNSplit) >= 2 && resourceARNSplit[0] == "log-group" {
+			return resourceARNSplit[1], arnParsed.Region, nil
+		}
+	}
+	return "", "", errors.Errorf("cannot get log group name from log group ARN: %s", logGroupARN)
+}
+
+func (in *awsCloudWatchInput) getLogEventsFromCloudWatch(svc cloudwatchlogsiface.ClientAPI) error {
+	errC := make(chan error, 1)
+	cwCtx := &cwContext{
+		refs: 1,
+		errC: errC,
+	}
+	defer cwCtx.done()
```

Comment:
Is `cwCtx` used at all? This was used in the context on e2e ACKs, but we don't have that here (we don't keep a record of our position in the stream to restart from there)

Reply:
Yeah it's only used for e2e ACK but we don't record the position... I should remove this.